### DISCUSSION TITLE
post/introducing-naudio-net-audio-toolkit

### DISCUSSION THREAD

**Discussion by giscus** [markheath/soundcode-blog-comments#59] [Announcements]
# post/introducing-naudio-net-audio-toolkit

Mark Heath's Development Blog

https://www.markheath.net/post/introducing-naudio-net-audio-toolkit

**Comment by Highwray** [comment]
Dear Sirs,
I am an amateur using C# and have developed a program to help with scales while learning to play the piano.  I would like to develop an addition so that a note appears on a stave display and, upon my ‘hitting’ a note, for the program to hear if it’s the correct one.  I have spent a few weeks scouring the internet for some clues as to how this could be done and NAudio seems to be close.  I understand that the use of GoertzelFilter only tells me if a specified frequency has been heard. This doesn’t seem successful unless the piano is perfect in producing the correct frequency (eg 440 for ‘A’).  If the frequency happened to be 445 or anything other than 440 then its not going to be successful. Is there a reasonable method within NAudio to give me the frequencies heard, which can then be compared to a range of values around the 440 mark for ‘A’?
Any help pointing me in the right direction would be much appreciated.
Michael

**Comment by Highwray** [comment]
Dear Sirs,
I am an amateur using C# and have developed a program to help with scales while learning to play the piano.  I would like to develop an addition so that a note appears on a stave display and, upon my ‘hitting’ a note, for the program to hear if it’s the correct one.  I have spent a few weeks scouring the internet for some clues as to how this could be done and NAudio seems to be close.  I understand that the use of GoertzelFilter only tells me if a specified frequency has been heard. This doesn’t seem successful unless the piano is perfect in producing the correct frequency (eg 440 for ‘A’).  If the frequency happened to be 445 or anything other than 440 then its not going to be successful. Is there a reasonable method within NAudio to give me the frequencies heard, which can then be compared to a range of values around the 440 mark for ‘A’?
Any help pointing me in the right direction would be much appreciated.
Michael

**Reply by Highwray** [reply]
Sorry I didn't realise quite how this works.  Old age at 80+ is finally settling in.

**Reply by markheath** [reply]
I suggest looking into algorithms used in guitar tuners. Autocorrelation of some form is quite commonly used - e.g. https://www.cycfi.com/2018/03/fast-and-efficient-pitch-detection-bitstream-autocorrelation/

**Comment by Highwray** [comment]
Thank you for your advice.  I will investigate and see if I can get something that I can understand.
Thanks again
Michael
